kymatio/scattering2d/tests/test_torch_scattering2d.py

```diff
@@ -0,0 +1,417 @@
+            S(x)
+        assert('Tensor must be of spatial size' in record.value.args[0])
+
+        S = Scattering2D(3, (32, 32), pre_pad=True, backend=backend, frontend='torch')
+
+        with pytest.raises(RuntimeError) as record:
+            S(x)
+        assert('Padded tensor must be of spatial size' in record.value.args[0])
+
+        x = torch.randn(8,8)
+        S = Scattering2D(2, (8, 8), backend=backend, frontend='torch')
+
+
+        x = x.to(device)
+        S = S.to(device)
+        if not (device == 'cpu' and backend.name == 'torch_skcuda'):
+            y = S(x)
+            assert(x.device == y.device)
+
+    # Check that several input size works
+    @pytest.mark.parametrize("backend", backends)
+    def test_input_size_agnostic(self, backend):
+        for N in [31, 32, 33]:
+            for J in [1, 2, 4]:
+                scattering = Scattering2D(J, shape=(N, N), backend=backend, frontend='torch')
+                x = torch.zeros(3, 3, N, N)
+
+                if backend.name == 'torch_skcuda':
+                    x = x.cuda()
+                    scattering.cuda()
+
+                S = scattering(x)
+                scattering = Scattering2D(J, shape=(N, N), pre_pad=True, backend=backend, frontend='torch')
+                x = torch.zeros(3,3,scattering.M_padded, scattering.N_padded)
+
+                if backend.name == 'torch_skcuda':
+                    x = x.cuda()
+                    scattering.cuda()
+
+                S = scattering(x)
+
+        N = 32
+        J = 5
+        scattering = Scattering2D(J, shape=(N, N), backend=backend, frontend='torch')
+        x = torch.zeros(3, 3, N, N)
+
+        if backend.name == 'torch_skcuda':
+            x = x.cuda()
+            scattering.cuda()
+
+        S = scattering(x)
+        assert(S.shape[-2:] == (1, 1))
+
+        N = 32
+        J = 5
+        scattering = Scattering2D(J, shape=(N+5, N), backend=backend, frontend='torch')
+        x = torch.zeros(3, 3, N+5, N)
+
+        if backend.name == 'torch_skcuda':
+            x = x.cuda()
+            scattering.cuda()
+
+        S = scattering(x)
+        assert (S.shape[-2:] == (1, 1))
+
+    def test_inputs(self):
+        with pytest.raises(ImportError) as ve:
+            scattering = Scattering2D(2, shape=(10, 10), frontend='torch', backend=fake_backend)
+        assert 'not supported' in ve.value.args[0]
+
+        with pytest.raises(RuntimeError) as ve:
+            scattering = Scattering2D(10, shape=(10, 10), frontend='torch')
+        assert 'smallest dimension' in ve.value.args[0]
+
+    @pytest.mark.parametrize("backend", backends)
+    @pytest.mark.parametrize("device", devices)
+    def test_gradients(self, backend, device):
+        if backend.name == 'torch_skcuda':
+            pytest.skip("The gradients are currently not implemented with the skcuda backend.")
+        else:
+            scattering = Scattering2D(2, shape=(8, 8), backend=backend, frontend='torch').double().to(device)
```

Comment:
Do all GPUs support double precision? Otherwise, this may fail on those.

Reply:
Yes, it's linked to torch mainly.

Reply:
Ok. Just checking.